scripts/gcs_upload.py

```diff
@@ -0,0 +1,578 @@
+        )
+    job_id = os.environ[ev]
+    if job_id == "":
+        raise EnvironmentVariableError("SLURM_JOB_ID is an empty string")
+    # gsutil writes temporary data to this directory
+    os.environ["TMPDIR"] = f"/scratch/fast/{job_id}/gsutil-tmpdir"
+
+
+async def _rename_blob(bucket, blob, target_path):
+    """
+    Rename the blob to target_path
+    Args:
+        bucket (google.cloud.storage.Bucket) the bucket instance
+        blob (google.cloud.storage.Blob) the blob instance
+        target_path (str) the new name for the blob
+    """
+    try:
+        bucket.rename_blob(blob, target_path)
+    except google.api_core.exceptions.NotFound:
+        # This seems to happen if the target blob already exists??
+        pass
+
+
+async def _rename_blobs(files, bucket_name, gcs_path, cloud_paths_map):
+    """
+    Args:
+        files (list): list of local filepaths that were uploaded
+        bucket_name (str): the name of the bucket where the files were uploaded
+        gcs_path (str): the cloud storage path where files were uploaded
+        cloud_paths_map (dict): a dictionary mapping uploaded file names (not paths)
+                                to their desired cloud paths
+    """
+    # convert flat list of blobs to original directory structure
+    # each rename operation is blocking, so fire all the http requests asynchronously then wait
+    storage_client = create_client()
+    bucket = storage_client.get_bucket(bucket_name)
+    tasks = []
+    for f in files:
+        name = PurePath(f).name
+        blob = bucket.get_blob(gcs_path + "/" + name)
+        target_path = cloud_paths_map[name]
+        if blob.name == target_path:
+            continue
+        logger.info(f"renaming blob from {blob.name} to {target_path}")
+        tasks.append(
+            _rename_blob(bucket=bucket, blob=blob, target_path=target_path)
+        )
+    await asyncio.gather(*tasks)
+
+
+def _gsutil_upload_worker(
+    bucket_name,
+    files,
+    gcs_path,
+    cloud_paths_map,
+    worker_id,
+    boto_options=None,
+):
+    """
+    Uploads a list of filepaths to GCS using gsutil
+    Args:
+        bucket_name (str): name of bucket to upload to
+        files (list): list of absolute filepaths to upload to the bucket
+        gcs_path (str): the cloud storage path to upload the files to
+        cloud_paths_map (dict): a dictionary mapping uploaded file names
+                                (not paths) to their target cloud paths
+        worker_id (int): A unique ID for this worker
+        boto_options (list): list of Boto options to pass to gsutil
+    """
+    options_str = ""
+    if boto_options is not None:
+        sep = " -o "
+        options_str = f"{sep}{sep.join(boto_options)}"
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        logfile = os.path.join(tmpdir, f"gsutil-cp-worker-{worker_id}-log.log")
+        catfile = os.path.join(tmpdir, f"file-subset-{worker_id}.txt")
```

Comment:
Do you need to route the list of files through an actual file? I feel like you should be able to do this via the `input` argument of `subprocess.run`. https://stackoverflow.com/a/48752179

Reply:
I see, will try that

Reply:
Addressed with https://github.com/AllenNeuralDynamics/nd-data-transfer/pull/6/commits/c141fb0104630733d07a8be850ca5d1bb1af058e